<!DOCTYPE html>
<html lang="en-US" dir="ltr">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <!--  
    Document Title
    =============================================
    -->
    <title>Kaleidoscope</title>
    <!--  
    Favicons
    =============================================
    -->
    <link rel="manifest" href="/manifest.json">
    <meta name="msapplication-TileColor" content="#ffffff">
    <meta name="theme-color" content="#ffffff">
    <!--  
    Stylesheets
    =============================================
    
    -->
    <!-- Default stylesheets-->
    <link href="assets/lib/bootstrap/dist/css/bootstrap.min.css" rel="stylesheet">
    <!-- Template specific stylesheets-->
    <link href="https://fonts.googleapis.com/css?family=Roboto+Condensed:400,700" rel="stylesheet">
    <link href="https://fonts.googleapis.com/css?family=Volkhov:400i" rel="stylesheet">
    <link href="https://fonts.googleapis.com/css?family=Open+Sans:300,400,600,700,800" rel="stylesheet">
    <link href="assets/lib/animate.css/animate.css" rel="stylesheet">
    <link href="assets/lib/components-font-awesome/css/font-awesome.min.css" rel="stylesheet">
    <link href="assets/lib/et-line-font/et-line-font.css" rel="stylesheet">
    <link href="assets/lib/flexslider/flexslider.css" rel="stylesheet">
    <link href="assets/lib/owl.carousel/dist/assets/owl.carousel.min.css" rel="stylesheet">
    <link href="assets/lib/owl.carousel/dist/assets/owl.theme.default.min.css" rel="stylesheet">
    <link href="assets/lib/magnific-popup/dist/magnific-popup.css" rel="stylesheet">
    <link href="assets/lib/simple-text-rotator/simpletextrotator.css" rel="stylesheet">
    <!-- Main stylesheet and color file-->
    <link href="assets/css/style.css" rel="stylesheet">
  </head>
  <body data-spy="scroll" data-target=".onpage-navigation" data-offset="10">
    <main>
      <div class="row no-gutter"></div>
      <div class="page-loader">
        <div class="loader">Loading...</div>
      </div>
      <!-- 
      <nav class="navbar navbar-custom navbar-transparent navbar-fixed-top one-page" role="navigation">
        <div class="container">
          <div class="navbar-header">
            <button class="navbar-toggle" type="button" data-toggle="collapse" data-target="#custom-collapse"><span class="sr-only">Toggle navigation</span><span class="icon-bar"></span><span class="icon-bar"></span><span class="icon-bar"></span></button><a class="navbar-brand" href="#"><img src="assets/images/ks_new_logo.png" width="40" style="margin-top:-10px"></a>
          </div>
          <div class="collapse navbar-collapse" id="custom-collapse">
            <ul class="nav navbar-nav navbar-right">
              <li><a href="#totop">Home</a></li>
              <li><a class="section-scroll" href="#app">App</a></li>
              <li><a class="section-scroll" href="#contact">Contact</a></li>
            </ul>
          </div>
        </div>
      </nav>
      -->
      <section class="home-section home-full-height bg-dark bg-gradient" id="home">
        <div class="titan-caption">
          <div class="caption-content">
            <img src="assets/images/ks_new_logo.svg" width="150px" height="150px">
            <div class="font-alt mb-40 titan-title-size-4">Kaleidoscope</div>
            <div class="font-alt mb-30 titan-title-size-1">In-Silico Data Exploration of Biological Datasets</div><a class="section-scroll btn btn-border-w btn-round" onclick="$('li:eq(1) a').tab('show');">Start Now</a>
          </div>
        </div>
      </section>
      <div class="main">
        <!--
        <section class="module" id="workflow">
          <div class="container">
            <div class="row">
              <div class="col-sm-6 col-sm-offset-3">
                <h2 class="module-title font-alt">About</h2>
                <div class="module-subtitle font-serif">Query your target genes across different biological databases and tools to explore your research questions in silico.</div>
                <img src="assets/images/KS_WorkflowFigure.png" alt="ks_workflow_figure" width="100%" height="100%"  /> 
              </div>
              
              <div class="col-md-5 col-md-push-7">
    <img src="assets/images/KS_WorkflowFigure.png" width="100%" height="100%" >
  </div>
  <div class="col-sm-6 col-md-pull-5">
    <p class="module-subtitle font-serif"> <b><big>Kaleidoscope</big></b> is an interactive R Shiny web application that provides a platform for an easy access to biological databases and bioinformatics tools via a user-friendly interface that could be explored by researchers to test hypothesis in silico. 
    The observations obtained from this platform could supplement existing hypotheses, spawn new ones and possibly direct future studies.</p>
  </div>
              
            </div>
          </div>
        </section> -->
        
        <section class="module pt-0 pb-0" id="about">
          <div class="row position-relative m-0">
            
            <div class="col-xs-12 col-md-6 side-image-text">
              <div class="row">
                <div class="col-sm-12">
                  <h2 class="module-title font-alt align-left">About</h2>
                  <div class="module-subtitle font-serif align-left">Query your target genes across different biological databases and tools to explore your research questions in silico.</div>
                  <p><b><big>Kaleidoscope</big></b> is an interactive R Shiny web application that provides a platform for an easy access to biological databases and bioinformatics tools via a user-friendly interface that could be explored by researchers to test hypothesis in silico.</p>
                  <p>The observations obtained from this platform could supplement existing hypotheses, spawn new ones and possibly direct future studies.</p>
                </div>
              </div>
            </div>
            
            <div class="col-xs-12 col-md-6 col-md-offset-6 side-image" style="display: flex; justify-content: center;">
              <img src="assets/images/KS_WorkflowFigure.png" style="margin-top: 125px; margin-bottom: 125px;" >
            </div>
            
          </div>
        </section>
        
        <hr class="divider-w">
        
        <section class="module" id="services">
          <div class="container">
            <div class="row">
              <div class="col-sm-6 col-sm-offset-3">
                <h2 class="module-title font-alt">Databases</h2>
                <div class="module-subtitle font-serif">Kaleidoscope is a pipeline app for data exploration and visualization that connects multiple databases and tools </div>
              </div>
            </div>
            <div class="row multi-columns-row">
              <div class="col-md-3 col-sm-6 col-xs-12">
               <!-- <a href="https://string-db.org/" target="_blank" rel="noopener noreferrer"> -->
                <div class="features-item">
                  <div class="features-icon"><span class="icon-genius"></span></div>
                  <h3 class="features-title font-alt">STRING</h3>
                  <p>Database and web resource of known and predicted protein–protein interactions.</p>
                </div>
                <!--</a> -->
              </div>
              <div class="col-md-3 col-sm-6 col-xs-12">
                <div class="features-item">
                  <div class="features-icon"><span class="icon-profile-male"></span></div>
                  <h3 class="features-title font-alt">GTEx</h3>
                  <p>Comprehensive public resource to study tissue-specific gene expression and regulation.</p>
                </div>
              </div>
              <div class="col-md-3 col-sm-6 col-xs-12">
                <div class="features-item">
                  <div class="features-icon"><span class="icon-linegraph"></span></div>
                  <h3 class="features-title font-alt">BrainCloud</h3>
                  <p>Patterns of gene expression in the brain across the lifespan of healthy humans</p>
                </div>
              </div>
              
              <div class="col-md-3 col-sm-6 col-xs-12">
                <div class="features-item">
                  <div class="features-icon"><span class="icon-bargraph"></span></div>
                  <h3 class="features-title font-alt">BRAIN RNA-Seq</h3>
                  <p>Gene expression of isolated and purified cells from human and mouse cortical tissue.</p>
                </div>
              </div>
              
              <div class="col-md-3 col-sm-6 col-xs-12">
                <div class="features-item">
                  <div class="features-icon"><span class="icon-lifesaver"></span></div>
                  <h3 class="features-title font-alt">iLINCS</h3>
                  <p>Genetic and chemical perturbed gene expression signatures</p>
                </div>
              </div>
              
              <div class="col-md-3 col-sm-6 col-xs-12">
                <div class="features-item">
                  <div class="features-icon"><span class="icon-map"></span></div>
                  <h3 class="features-title font-alt">GWAS Catalog</h3>
                  <p>Largest curated collection of all published genome-wide association studies.</p>
                </div>
              </div>
              <div class="col-md-3 col-sm-6 col-xs-12">
                <div class="features-item">
                  <div class="features-icon"><span class="icon-search"></span></div>
                  <h3 class="features-title font-alt">Lookup</h3>
                  <p>Differential gene expression signatures across different diseases and drug treatments</p>
                </div>
              </div>
              <div class="col-md-3 col-sm-6 col-xs-12">
                <div class="features-item">
                  <div class="features-icon"><span class="icon-aperture"></span></div>
                  <h3 class="features-title font-alt">BrainAtlas</h3>
                  <p>Gene expression of nuclei derived from frozen human brain specimens to survey cell type diversity</p>
                </div>
              </div>
            </div>
          </div>
        </section>


        <hr class="divider-w">
        
        <section class="module bg-dark-60">
          <div class="container">
            <div class="row">
              <div class="col-sm-6 col-sm-offset-3">
                <h2 class="module-title font-alt">Statistics</h2>
                <div class="module-subtitle font-serif"></div>
              </div>
            </div>
            <div class="row multi-columns-row">
              <div class="col-sm-6 col-md-3 col-lg-3">
                <div class="count-item mb-sm-40">
                  <div class="count-icon"><span class="icon-map-pin"></span></div>
                  <h3 class="count-to font-alt" data-countto="4319591"></h3>
                  <h5 class="count-title font-serif">Datapoints in the lookup datasets</h5>
                </div>
              </div>
              <div class="col-sm-6 col-md-3 col-lg-3">
                <div class="count-item mb-sm-40">
                  <div class="count-icon"><span class="fa fa-database"></span></div>
                  <h3 class="count-to font-alt" data-countto="8"></h3>
                  <h5 class="count-title font-serif">Major Biological Databases</h5>
                </div>
              </div>
              <div class="col-sm-6 col-md-3 col-lg-3">
                <div class="count-item mb-sm-40">
                  <div class="count-icon"><span class="icon-tools-2"></span></div>
                  <h3 class="count-to font-alt" data-countto="5"></h3>
                  <h5 class="count-title font-serif">APIs</h5>
                </div>
              </div>
              <div class="col-sm-6 col-md-3 col-lg-3">
                <div class="count-item mb-sm-40">
                  <div class="count-icon"><span class="icon-upload"></span></div>
                  <h3 class="count-to font-alt" data-countto="266"></h3>
                  <h5 class="count-title font-serif">Lookup Replication Datasets</h5>
                </div>
              </div>
            </div>
          </div>
        </section>
        
        <section class="module" id="pubs">
          <div class="container">
            <div class="row">
              <div class="col-sm-12">
                <h2 class="module-title font-alt">Publications</h2>
                <div class="module-subtitle font-serif">Kaleidoscope was used in several publications, including:</div>
                <div>
                  <p class="lead" style= "font-size: x-large; font-weight: bold"><a data-toggle="collapse" data-pare="#accordionThree" href="#collapseOne">
                    Connectivity Analyses of Bioenergetic Changes in Schizophrenia: Identification of Novel Treatments </a></p>
                  <p style="font-size: small; font-style: italic; text-decoration: underline">Molecular Neurobiology</p>
                  <p style = "font-size: small">Courtney Sullivan, Catharine Mielnik, Sinead O'Donovan, Adam Funk, Eduard Bentea, Erica A DePasquale, Khaled Alganem, Zhexing Wen, Vahram Haroutunian, Pavel Katsel, Amy Ramsey, Jarek Meller, Robert E McCullumsmith</p>
                  <div class="panel-group transparent-panel-group" id="accordionThree">
                    <div class="panel panel-default">
                      <div id="collapseOne" class="panel-collapse collapse">
                        <div class="panel-body text-left">
                          <p>We utilized a cell-level approach to examine glycolytic pathways in the DLPFC of subjects with schizophrenia (n = 16) and control (n = 16) and found decreased mRNA expression of glycolytic enzymes in pyramidal neurons, but not astrocytes. To replicate these novel bioenergetic findings, we probed independent datasets for bioenergetic targets and found similar abnormalities. Next, we used a novel strategy to build a schizophrenia bioenergetic profile by a tailored application of the Library of Integrated Network-Based Cellular Signatures data portal (iLINCS) and investigated connected cellular pathways, kinases, and transcription factors using Enrichr. Finally, with the goal of identifying drugs capable of "reversing" the bioenergetic schizophrenia signature, we performed a connectivity analysis with iLINCS and identified peroxisome proliferator-activated receptor (PPAR) agonists as promising therapeutic targets. We administered a PPAR agonist to the GluN1 knockdown model of schizophrenia and found it improved long-term memory. Taken together, our findings suggest that tailored bioinformatics approaches, coupled with the LINCS library of transcriptional signatures of chemical and genetic perturbagens, may be employed to identify novel treatment strategies for schizophrenia and related diseases. <a href="https://pubmed.ncbi.nlm.nih.gov/30338483/" target="_blank" rel="noopener noreferrer" style="font-size: medium;">PMID: 30338483</a></p>

                        </div>
                      </div>
                   </div>
                </div>
                
                <p class="lead" style= "font-size: x-large; font-weight: bold"><a data-toggle="collapse" href="#collapseTwo" aria-expanded="false" aria-controls="collapseTwo" >
                    A bioinformatic inquiry of the EAAT2 interactome in postmortem and neuropsychiatric datasets </a></p>
                  <p style="font-size: small; font-style: italic; text-decoration: underline">Schizophrenia Research</p>
                  <p style = "font-size: small">Sophie Asah, Khaled Alganem, Robert McCullumsmith, Sinead O'Donovan</p>
                  <div class="panel-group transparent-panel-group" id="accordionTwo">
                    <div class="panel panel-default">
                      <div id="collapseTwo" class="panel-collapse collapse">
                        <div class="panel-body text-left">
                          <p>Altered expression and localization of the glutamate transporter EAAT2 is found in schizophrenia and other neuropsychiatric (major depression, MDD) and neurological disorders (amyotrophic lateral sclerosis, ALS). However, the EAAT2 interactome, the network of proteins that physically or functionally interact with EAAT2 to support its activity, has yet to be characterized in severe mental illness. We compiled a list of "core" EAAT2 interacting proteins. Using Kaleidoscope, an R-shiny application, we data mined publically available postmortem transcriptome datasets to determine whether components of the EAAT2 interactome are differentially expressed in schizophrenia and, using Reactome, identify which interactome-associated biological pathways are altered. Overall, these "look up" studies highlight region-specific, primarily frontal cortex (dorsolateral prefrontal cortex and anterior cingulate cortex), changes in the EAAT2 interactome and implicate altered metabolism pathways in schizophrenia. Pathway analyses also suggest that perturbation of components of the EAAT2 interactome in animal models of antipsychotic administration impact metabolism. Similar changes in metabolism pathways are seen in ALS, in addition to altered expression of many components of the EAAT2 interactome. However, although EAAT2 expression is altered in a postmortem MDD dataset, few other components of the EAAT2 interactome are changed. Thus, "look up" studies suggest region- and disease-relevant biological pathways related to the EAAT2 interactome that implicate glutamate reuptake perturbations in schizophrenia, while providing a useful tool to exploit "omics" datasets. <a href="https://pubmed.ncbi.nlm.nih.gov/32197935/" target="_blank" rel="noopener noreferrer" style="font-size: medium;">PMID: 32197935</a></p>

                        </div>
                      </div>
                   </div>
                </div>
                
                <p class="lead" style= "font-size: x-large; font-weight: bold"><a data-toggle="collapse" data-pare="#accordionThree" href="#collapseThree">
                    Single molecule in situ hybridization reveals distinct localizations of schizophrenia risk-related transcripts SNX19 and AS3MT in human brain </a></p>
                  <p style="font-size: small; font-style: italic; text-decoration: underline">Molecular Psychiatry</p>
                  <p style = "font-size: small">Yoichiro Takahashi, Kristen R. Maynard, Madhavi Tippani, Andrew E. Jaffe, Keri Martinowich, Joel E. Kleinman, Daniel R. Weinberger & Thomas M. Hyde</p>
                  <div class="panel-group transparent-panel-group" id="accordionThree">
                    <div class="panel panel-default">
                      <div id="collapseThree" class="panel-collapse collapse">
                        <div class="panel-body text-left">
                          <p>Genome-wide association studies have identified single nucleotide polymorphisms (SNPs) associated with schizophrenia risk. Integration of RNA-sequencing data from postmortem human brains with these risk SNPs identified transcripts associated with increased schizophrenia susceptibility, including a class of exon 9-spliced isoforms of Sorting nexin-19 (SNX19d9) and an isoform of Arsenic methyltransferase (AS3MT) splicing out exons 2 and 3 (AS3MTd2d3). However, the biological function of these transcript variants is unclear. Defining the cell types where these risk transcripts are dominantly expressed is an important step to understand function, in prioritizing specific cell types and/or neural pathways in subsequent studies. To identify the cell type-specific localization of SNX19 and AS3MT in the human dorsolateral prefrontal cortex (DLPFC), we used single-molecule in situ hybridization techniques combined with automated quantification and machine learning approaches to analyze 10 postmortem brains of neurotypical individuals. These analyses revealed that both pan-SNX19 and pan-AS3MT were more highly expressed in neurons than non-neurons in layers II/III and VI of DLPFC. Furthermore, pan-SNX19 was preferentially expressed in glutamatergic neurons, while pan-AS3MT was preferentially expressed in GABAergic neurons. Finally, we utilized duplex BaseScope technology, to delineate the localization of SNX19d9 and AS3MTd2d3 splice variants, revealing consistent trends in spatial gene expression among pan-transcripts and schizophrenia risk-related transcript variants. These findings demonstrate that schizophrenia risk transcripts have distinct localization patterns in the healthy human brains, and suggest that SNX19 transcripts might disrupt the normal function of glutamatergic neurons, while AS3MT may lead to disturbances in the GABAergic system in the pathophysiology of schizophrenia. <a href="https://pubmed.ncbi.nlm.nih.gov/33649454/" target="_blank" rel="noopener noreferrer" style="font-size: medium;">PMID: 33649454</a></p>

                        </div>
                      </div>
                   </div>
                </div>
                
                <p class="lead" style= "font-size: x-large; font-weight: bold"><a data-toggle="collapse" data-pare="#accordionThree" href="#collapseFour">
                    Role of Astrocytes in Major Neuropsychiatric Disorders </a></p>
                  <p style="font-size: small; font-style: italic; text-decoration: underline">Neurochemical Research </p>
                  <p style = "font-size: small">Xiaolu Zhang, Rawan S. Alnafisah, Abdul-Rizaq A. Hamoud, Rammohan Shukla, Zhexing Wen, Robert E. McCullumsmith & Sinead M. O’Donovan</p>
                  <div class="panel-group transparent-panel-group" id="accordionThree">
                    <div class="panel panel-default">
                      <div id="collapseFour" class="panel-collapse collapse">
                        <div class="panel-body text-left">
                          <p>Astrocytes are the primary homeostatic cells of the central nervous system, essential for normal neuronal development and function, metabolism and response to injury and inflammation. Here, we review postmortem studies examining changes in astrocytes in subjects diagnosed with the neuropsychiatric disorders schizophrenia (SCZ), major depressive disorder (MDD), and bipolar disorder (BPD). We discuss the astrocyte-related changes described in the brain in these disorders and the potential effects of psychotropic medication on these findings. Finally, we describe emerging tools that can be used to study the role of astrocytes in neuropsychiatric illness. <a href="https://pubmed.ncbi.nlm.nih.gov/33411227/" target="_blank" rel="noopener noreferrer" style="font-size: medium;">PMID: 33411227</a></p>

                        </div>
                      </div>
                   </div>
                </div>
                
                <p class="lead" style= "font-size: x-large; font-weight: bold"><a data-toggle="collapse" data-pare="#accordionThree" href="#collapseFive">
                    Markers of copper transport in the cingulum bundle in schizophrenia </a></p>
                  <p style="font-size: small; font-style: italic; text-decoration: underline">Schizophrenia Research </p>
                  <p style = "font-size: small">Kirsten E Schoonover, Rosalinda C Roberts</p>
                  <div class="panel-group transparent-panel-group" id="accordionThree">
                    <div class="panel panel-default">
                      <div id="collapseFive" class="panel-collapse collapse">
                        <div class="panel-body text-left">
                          <p>Imaging and postmortem studies indicate that schizophrenia subjects exhibit abnormal connectivity in several white matter tracts, including the cingulum bundle. Copper chelators given to experimental animals damage myelin and myelin-producing oligodendrocytes, and the substantia nigra of schizophrenia subjects shows lower levels of copper, copper transporters, and copper-utilizing enzymes. This study aimed to elucidate the potential role of copper homeostasis in white matter pathology in schizophrenia. Protein levels of the copper transporters ATP7A and CTR1, and dysbindin-1, an upstream modulator of copper metabolism and schizophrenia susceptibility factor, were measured using Western blot analyses of the postmortem cingulum bundle of schizophrenia subjects (n=16) and matched controls (n=13). Additionally, the patient group was subdivided by treatment status: off- (n=8) or on-medication (n=8). Relationships between proteins from the current study were correlated among themselves and markers of axonal integrity previously measured in the same cohort. Schizophrenia subjects exhibited similar protein levels to controls, with no effect of antipsychotic treatment. The dysbindin-1A/1BC relationship was positive in controls and schizophrenia subjects; however, antipsychotic treatment appeared to reverse this relationship in a statistically different manner from that of controls and unmedicated subjects. The relationships between dysbindin-1A/neurofilament heavy and ATP7A/α-tubulin were positively correlated in the schizophrenia group that was significantly different from the lack of correlation in controls. Copper transporters and dysbindin-1 appear to be more significantly affected in the grey matter of schizophrenia subjects. However, the relationships among proteins in white matter may be more substantial and dependent on treatment status. <a href="https://pubmed.ncbi.nlm.nih.gov/33434726/" target="_blank" rel="noopener noreferrer" style="font-size: medium;">PMID: 33434726</a></p>

                        </div>
                      </div>
                   </div>
                </div>

              </div>
              
              
              
            </div>

          </div>
          </div>
        </section>


        <section class="module-small bg-dark-60">
          <div class="container">
            <div class="row">
              <div class="col-sm-6">
                <div class="widget">
                  <h5 class="widget-title font-alt">Developed By</h5>
                  <p>
                    Khaled Alganem<br>
                    Cognitive Disorders Research Laboratory (CDRL)<br>
                    University of Toledo College of Medicine and Life Sciences<br>
                    Toledo, Ohio<br>
                    khaled.alganem@rockets.toledo.edu
                  </p>
                  <p>
                    <a href="https://cdrl-ut.org/" target="_blank" rel="noopener noreferrer"><i class="fa fa-dribbble" style="font-size:20px; padding-right:15px"></i></a>
                    <a href="https://github.com/kalganem" target="_blank" rel="noopener noreferrer"><i class="fa fa-github" style="font-size:20px; padding-right:25px"></i></a>
                  </p>
                </div>
              </div>
              <div class="col-sm-6">

              </div>
            </div>
          </div>
        </section>  

      </div>
      <div class="scroll-up"><a href="#totop"><i class="fa fa-angle-double-up"></i></a></div>

    </main>
    <!--  
    JavaScripts
    =============================================
    -->
    <!--  
    <script src="assets/lib/jquery/dist/jquery.js"></script>
    
    <script src="assets/lib/bootstrap/dist/js/bootstrap.min.js"></script>
    -->
    <script src="assets/lib/wow/dist/wow.js"></script>
    <script src="assets/lib/isotope/dist/isotope.pkgd.js"></script>
    <script src="assets/lib/imagesloaded/imagesloaded.pkgd.js"></script>
    <script src="assets/lib/magnific-popup/dist/jquery.magnific-popup.js"></script>
    <script src="assets/lib/simple-text-rotator/jquery.simple-text-rotator.min.js"></script>
    <script src="assets/lib/masonry/dist/masonry.pkgd.js"></script>
    <script src="assets/lib/jquery.mb.ytplayer/dist/jquery.mb.YTPlayer.js"></script>
    <script src="assets/js/plugins.js"></script>
    <script src="assets/js/main.js"></script>
  </body>
</html>